plugins/woocommerce/src/Blocks/BlockTypes/ProductSummary.php

```diff
@@ -44,11 +48,198 @@ protected function get_block_type_supports() {
+			case 'characters_including_spaces':
+				return strlen( $text );
+			case 'words':
+			default:
+				return str_word_count( wp_strip_all_tags( $text ) );
+		}
+	}
+
+	/**
+	 * Trim characters to a specified length.
+	 *
+	 * @param string $text Text to trim.
+	 * @param int    $max_length Maximum length of the text.
+	 * @param string $count_type What is being counted. One of 'words', 'characters_excluding_spaces', or 'characters_including_spaces'.
+	 * @return string Trimmed text.
+	 */
+	private function trim_characters( $text, $max_length, $count_type ) {
+		$pure_text = wp_strip_all_tags( $text );
+		$trimmed   = mb_substr( $pure_text, 0, $max_length );
+
+		if ( 'characters_including_spaces' === $count_type ) {
+			return $trimmed;
+		}
+
+		preg_match_all( '/([\s]+)/', $trimmed, $spaces );
+		$space_count = ! empty( $spaces[0] ) ? count( $spaces[0] ) : 0;
+		return mb_substr( $pure_text, 0, $max_length + $space_count );
+	}
+
+	/**
+	 * Generates the summary text from a string of text. It's not ideal
+	 * but allows keeping the editor and frontend consistent.
+	 *
+	 * NOTE: If editing, keep it in sync with generateSummary function from
+	 * plugins/woocommerce-blocks/assets/js/base/components/summary/utils.ts!
+	 *
+	 * Once HTML API allow for HTML manipulation both functions (PHP and JS)
+	 * should be updated to solution fully respecting the word count.
+	 * https://github.com/woocommerce/woocommerce/issues/52835
+	 *
+	 * @param string $source     Source text.
+	 * @param int    $max_length Limit number of items returned if text has multiple paragraphs.
+	 * @return string Generated summary.
+	 */
+	private function generate_summary( $source, $max_length ) {
+		$count_type             = wp_get_word_count_type();
+		$source_with_paragraphs = wpautop( $source );
+		$source_word_count      = $this->count_text( $source_with_paragraphs, $count_type );
+
+		if ( $source_word_count <= $max_length ) {
+			return $source_with_paragraphs;
+		}
+
+		$first_paragraph            = $this->get_first_paragraph( $source_with_paragraphs );
+		$first_paragraph_word_count = $this->count_text( $first_paragraph, $count_type );
+
+		if ( $first_paragraph_word_count <= $max_length ) {
+			return $first_paragraph;
+		}
+
+		if ( 'words' === $count_type ) {
+			return wpautop( wp_trim_words( $first_paragraph, $max_length ) );
+		}
+
+		return $this->trim_characters( $first_paragraph, $max_length, $count_type ) . '…';
+	}
+
+	/**
+	 * Include and render the block.
+	 *
+	 * @param array    $attributes Block attributes. Default empty array.
+	 * @param string   $content    Block content. Default empty string.
+	 * @param WP_Block $block      Block instance.
+	 * @return string Rendered block type output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		if ( ! empty( $content ) ) {
+			parent::register_block_type_assets();
+			$this->register_chunk_translations( [ $this->block_name ] );
```

Comment:
Is this check necessary when the block is used in the All Products block and the JS version is loaded? Should we add an explicit comment about this?

Reply:
I see what you mean but it's exactly the same pattern across all product elements, for example:

https://github.com/woocommerce/woocommerce/blob/f2860f25debfe728fdd7b88734e1c11c1dbd127b/plugins/woocommerce/src/Blocks/BlockTypes/ProductImage.php#L200-L205

https://github.com/woocommerce/woocommerce/blob/f2860f25debfe728fdd7b88734e1c11c1dbd127b/plugins/woocommerce/src/Blocks/BlockTypes/ProductPrice.php#L84-L88

https://github.com/woocommerce/woocommerce/blob/f2860f25debfe728fdd7b88734e1c11c1dbd127b/plugins/woocommerce/src/Blocks/BlockTypes/ProductSKU.php#L51-L55

I guess, if anything,  it could be somehow abstracted to the class above but I don't think the comment in Product Summary is necessary in that case. 🙌